### ISSUE TITLE
Add support for scripted expression in Elasticsearch flow

### ISSUE THREAD

**Issue by T-Hazzle** [akka/alpakka#1879]
### Short description

Elasticsearch has broad support for scripting, in both query and indexing operations:

https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-scripting.html

I’d like to add support for scripting in alpakka as part of the Flow/Sink stage.

### Details

Certain operations are only possible through scripting in Elasticsearch.

In my case, I am working with a nested document structure. This blog post has details: https://iridakos.com/how-to/2019/05/02/add-update-delete-elasticsearch-nested-objects.html

The bulk API supports scripted updates. This example is based on what they show on their website:

```
POST _bulk
{ "update" : { "_id" : "00004", "_index" : "index1", "retry_on_conflict" : 3} }
{ "script" : { "source": "ctx._source.counter += params.amount", "lang" : "painless", "params" : {"param1" : 1}}, "upsert" : {"counter" : 1}}
{ "update" : { "_id" : "00004", "_index" : "index1", "retry_on_conflict" : 3} }
{ "script" : { "id": "increment", "lang" : "painless", "params" : {"param1" : 1}}, "upsert" : {"counter" : 1}}
```

Scripts can be inline, or reference a previously prepared script. Scripts can be written in painless, lucene expressions or use a custom scripting engine.

The API can also accept a document as an upsert.

I think we should split the creation into two stages to avoid a combinatorial explosion of factory methods. This would involve a new Script class with its own factories - this class can be used in the future for queries, as the structure is always the same in Elasticsearch. And a new factory in WriteMessage that accepts a script property.

So something like:
```
WriteMessage.createScriptedMessage(id = “0001”, script = Script.createPreparedScript(id = “increment”, params = Amount(1)))

WriteMessage.createScriptedMessage(id = “0001”, script = Script.createInlineScript(source = “ctx._source.counter += params.amount”, lang = “painless”, params = Amount(1)))
```

Add a new private function similar to this: https://github.com/akka/alpakka/blob/master/elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/impl/ElasticsearchFlowStage.scala#L211

But have it render a scripting message instead of a document based message.

Then have the map function here pick the right helper function to generate the JSON:
https://github.com/akka/alpakka/blob/master/elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/impl/ElasticsearchFlowStage.scala#L192

Let me know your thoughts. This is my first contribution and I’m more experienced with Java than Scala so I will likely need some feedback.

Regards,
Tom

**Comment by T-Hazzle** [comment]
I've signed the CLA.

Tom

**Comment by 2m** [comment]
Hi Tom. Extending functionality via a custom `WriteMessage` sounds good. Feel free to create a PR early in the process if you have any questions or need feedback!

**Comment by T-Hazzle** [comment]
Thanks Martynas, I'll get started.

Tom

**Comment by AndreaPicasso** [comment]
I'm also excited for this functionality. Where can i find the code or contribute?

**Comment by T-Hazzle** [comment]
Hi @AndreaPicasso 

I've push a fork and added you as a collaborator. I've also just raised a provisional PR.

Sorry for the delay, I got a bit lost trying to avoid making a breaking change to the API, but I think this is necessary to allow for both script params and upsert documents to be included as abstract types. So I've started again with that approach.

So long as the breaking change is acceptable, the next steps would be to update ElasticsearchFlowStage and update the DSLs.

Tom

**Comment by AndreaPicasso** [comment]
Ok great!
Thanks

**Comment by T-Hazzle** [comment]
I'm still working on this, I've pushed my latest efforts to my fork.

I'm quite stuck at the moment. I've been trying to make the script paramsan abstract object. It mess's with the existing mechanism used to serialise the messages via an implicit string message writer.

As a potential workaround, I may simplify the params to a hashmap and see if that makes things easier.

**Comment by fabboe** [comment]
Currently we POSTs to the API are treated differently by "Operation", likely that's what caused different `WriteMessage` constructors in the past. 

Supporting scripts actually isn't about supporting a new Operation (e.g. a new `WriteMessage` flavour) - it's about passing on the payload more flexibly (currently [hardcoded to "doc"](https://github.com/akka/alpakka/blob/master/elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/impl/RestBulkApi.scala#L47))

Have you considered exposing the `message.source` more raw in the API? e.g. expose the whole payload of the line after the "operation payload".

One way to do this backwards-compatible could be to introduce another parameter (e.g. `payloadType`) to `WriteMessage` that defaults to `doc`.

That way it's also potentially more forward compatible as one can control  the actual message payload. From the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#bulk-update):

> The update action payload supports the following options: doc (partial document), upsert, doc_as_upsert, script, params (for script), lang (for script), and _source. See update documentation for details on the options. 

Even better would be to expose the whole payload, so one can freely combine e.g. `doc` and `_source` or other options from the docs. That might be difficult to do backwards-compatible.

What do you think?

**Comment by ennru** [comment]
Did you have a look at the latest state in `master`? The implementation recently changed to use a raw Akka HTTP client (instead of ES REST client library) and that breaks the API compared to 2.0.x already.

**Comment by ennru** [comment]
@yasha-dev1 No, this discussion has been quiet for a long while now. Check out https://github.com/akka/alpakka/blob/master/CONTRIBUTING.md for our contribution hints. Welcome!
